compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -248,6 +248,138 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+     */
+    private def compileAsStaticMethod(sym: Symbol): Boolean = {
+      sym.isStaticMember || {
+        isStatifyCandidate(sym) &&
+        !statifyCandidateMethodsThatReferenceThis.contains(sym)
+      }
+    }
+
+    /** Finds all statify-candidate methods that reference `this`, inspired by Delambdafy. */
+    private final class ThisReferringMethodsTraverser extends Traverser {
+      // the set of statify-candidate methods that directly refer to `this`
+      private val roots = mutable.Set.empty[Symbol]
+
+      // for each statify-candidate method `m`, the set of statify-candidate methods that call `m`
+      private val methodReferences = mutable.Map.empty[Symbol, mutable.Set[Symbol]]
+
+      def methodReferencesThisIn(tree: Tree): collection.Set[Symbol] = {
+        traverse(tree)
+        propagateReferences()
+      }
+
+      private def addRoot(symbol: Symbol): Unit =
+        roots += symbol
+
+      private def addReference(from: Symbol, to: Symbol): Unit =
+        methodReferences.getOrElseUpdate(to, mutable.Set.empty) += from
+
+      private def propagateReferences(): collection.Set[Symbol] = {
+        val result = mutable.Set.empty[Symbol]
+
+        def rec(symbol: Symbol): Unit = {
+          // mark `symbol` as referring to `this`
+          if (result.add(symbol)) {
+            // `symbol` was not yet in the set; propagate further
+            for {
+              referrers <- methodReferences.get(symbol)
+              referrer <- referrers
+            } {
+              // referrer calls `symbol`, so it transitively references `this`
+              rec(referrer)
+            }
+          }
+        }
+
+        roots.foreach(rec(_))
+
+        result
+      }
+
+      private var currentMethod: Symbol = NoSymbol
+
+      override def traverse(tree: Tree): Unit = tree match {
+        case _: DefDef =>
+          if (isStatifyCandidate(tree.symbol)) {
+            currentMethod = tree.symbol
+            super.traverse(tree)
+            currentMethod = NoSymbol
+          } else {
+            // No need to traverse other methods; we always assume they refer to `this`.
+          }
+        case Function(_, Apply(target, _)) =>
+          /* We don't drill into functions because at this phase they will always refer to `this`.
+           * They will be of the form {(args...) => this.anonfun(args...)}
+           * but we do need to make note of the lifted body method in case it refers to `this`.
+           */
+          if (currentMethod.exists)
+            addReference(from = currentMethod, to = target.symbol)
+        case Apply(sel @ Select(This(_), _), args) if isStatifyCandidate(sel.symbol) =>
+          if (currentMethod.exists)
+            addReference(from = currentMethod, to = sel.symbol)
+          super.traverseTrees(args)
+        case This(_) =>
+          if (currentMethod.exists && tree.symbol == currentMethod.enclClass)
```

Comment:
What does it mean if `tree.symbol != currentMethod.enclClass`?

Reply:
It can apparently be a static module to be loaded with `genLoadModule`. I added a comment.